### ISSUE TITLE
[BUG]: v5.0.0-beta.12 seems not to gather from API

### ISSUE THREAD

**Issue by loewensteinph** [kamontat/fthelper#23]
# Bug report

## Describe the bug

> Prometheus does not scrape ftrqtrade related metrics

## To Reproduce

1. configure docker-compose file as following:

`  ftmetric1:
    image: ghcr.io/kamontat/ftmetric:v5.0.0-beta.12
    container_name: ftmetric1
    restart: on-failure
    ports:
      - 8091:8091
    environment:
      - FTH_FREQTRADE__DB__ENABLED=false
      - FTH_SERVER__PORT=8091
      - FTH_FREQTRADE__HTTP__APIVER=v1
      - FTH_FREQTRADE__HTTP__URL=http://freqtrade1:8081
      - FTH_FREQTRADE__HTTP__USERNAME=freqtrader
      - FTH_FREQTRADE__HTTP__PASSWORD=secret      
      - FTH_FREQTRADE__HTTP__ENABLED=true
    command: --no-env-file --debug
    depends_on:
      - freqtrade1`

## Expected behavior

> Result of freqtrade_mode non empty

## Information

Try to complete information as much as possible.

1. Desktop:
 - OS: Windows 11
 - WSL2 using docker
 - Browser firefox

## Additional context

http://127.0.0.1:9090/targets shows status up
http://127.0.0.1:8091/metrics not showing metrics related to freqtrade
postman shows result for http://127.0.0.1:8081/api/v1/profit

**Comment by kamontat** [comment]
Any error or warning log?

**Comment by loewensteinph** [comment]
Log looks fine - no warnings or errors.

**Comment by kamontat** [comment]
could you try run following command one line at the time.

```bash
docker exec -it ftmetric1 bash
curl freqtrade1:8080/api/v1/ping
```

It should show something..

**Comment by kamontat** [comment]
could you try run following command one line at the time.

```bash
docker exec -it ftmetric1 bash
curl freqtrade1:8080/api/v1/ping
```

It should show something..

**Comment by richardbeumer** [comment]
same issue here. When using Busybox image I see:  
```[DEBUG] | override 'freqtrade.http.url' from '<**** [<nil>] -> '"http:********************lab" [string] (env)```
```[DEBUG] | override 'freqtrade.url' from 'http:*************080 [string] -> 'http:*************080 [string] (env)```

However I only have set ```FTH_FREQTRADE__HTTP__URL``` variable
In my logs I see ```freqtrade:connection [DEBUG] | error: Get "http://freqtrade:8080/api/v1/logs?limit=100": dial tcp: lookup freqtrade on 127.0.0.11:53: no such host```

Apparently it is overwriting ```FTH_FREQTRADE__HTTP__URL``` variable with ``````FTH_FREQTRADE__URL``` variable

Please help

**Comment by frysee** [comment]
Same here:
`19:07:44 fs:type:resolver     [DEBUG] | mark /fth/configs as directory
19:07:44 config:builder       [DEBUG] | loading config from wrapper of 'multiple': [/fth/configs (directory)]
19:07:44 map:merger           [DEBUG] | merging key=clusters from <nil> -> []
19:07:44 map:merger           [DEBUG] | merging key=freqtrade from <nil> -> map[cache:map[balance:3m count:1m locks:5m logs:1m performance:1m ping:1s profit:5m schedule_perf:1m show_config:1m stats:5m version:10m whitelist:15m] db:map[enabled:true name:freqtrade password: type:postgres url:localhost:5432 username:] http:map[apiver:v1 enabled:true password: query:map[logs:map[limit:100]] url:http://localhost:8080 username:]]
19:07:44 map:merger           [DEBUG] | merging key=server from <nil> -> map[metrics-path:/metrics port:8090]
19:07:44 map:merger           [DEBUG] | merging key=warmup from <nil> -> map[enabled:true interval:30s]
19:07:44 config:builder       [DEBUG] | found some config in env
19:07:44 config:builder       [DEBUG] | override 'freqtrade.db.enabled' from 't*** [bool] -> 'f**** [bool] (env)
19:07:44 config:builder       [DEBUG] | override 'freqtrade.http.url' from 'http:*************080 [string] -> 'http:/*****************081 [string] (env)
19:07:44 config:builder       [DEBUG] | override 'freqtrade.http.username' from ' [string] -> 'fr******e [string] (env)
19:07:44 config:builder       [DEBUG] | override 'freqtrade.http.password' from ' [string] -> 'fr******e [string] (env)
19:07:44 command:default      [INFO ] | warmup process: false (interval=30s)
19:07:44 command:default      [INFO ] | Start server at 0.0.0.0:8090
19:08:09 cache:service        [DEBUG] | cannot get data key: warmup.success.rate
19:08:09 command:default      [INFO ] | skip 'internal_warmup' because success rate is nil
19:08:09 cache:service        [DEBUG] | cannot get data key: warmup.durations
19:09:09 cache:service        [DEBUG] | cannot get data key: warmup.success.rate
19:09:09 command:default      [INFO ] | skip 'internal_warmup' because success rate is nil
19:09:09 cache:service        [DEBUG] | cannot get data key: warmup.durations
19:10:09 cache:service        [DEBUG] | cannot get data key: warmup.success.rate
19:10:09 command:default      [INFO ] | skip 'internal_warmup' because success rate is nil
19:10:09 cache:service        [DEBUG] | cannot get data key: warmup.durations
19:11:09 cache:service        [DEBUG] | cannot get data key: warmup.success.rate
19:11:09 command:default      [INFO ] | skip 'internal_warmup' because success rate is nil
19:11:09 cache:service        [DEBUG] | cannot get data key: warmup.durations
19:12:09 cache:service        [DEBUG] | cannot get data key: warmup.success.rate
19:12:09 command:default      [INFO ] | skip 'internal_warmup' because success rate is nil
19:12:09 cache:service        [DEBUG] | cannot get data key: warmup.durations`

**Comment by loewensteinph** [comment]
Either i don't understand how it is supposed to work, or a simple one cluster env setup does not work:

console args "--no-env-file --debug"

    environment:
      - FTH_CLUSTERS=1A
      - FTH_CLUSTER__1A__DB__ENABLED=false      
      - FTH_CLUSTER__1A__HTTP__ENABLED=true
      - FTH_CLUSTER__1A__HTTP__URL="http://freqtrade1:8081"
      - FTH_CLUSTER__1A__HTTP__APIVER=v1
      - FTH_CLUSTER__1A__HTTP__USERNAME="freqtrader"
      - FTH_CLUSTER__1A__HTTP__PASSWORD="secret"
      - FTH_SERVER__PORT=8091

09:11:28 command:default      [INFO ] | [with-cache] Connection cluster '1A'  http     - http://localhost:8080 ':'  database - localhost:5432 (postgres) ':'
09:11:28 client:http:1A       [DEBUG] | request: GET http://localhost:8080/api/v1/ping
09:11:28 cache:service        [DEBUG] | creating 'http.total-call.1A' data
09:11:28 client:http:1A       [DEBUG] | error: Get "http://localhost:8080/api/v1/ping": dial tcp 127.0.0.1:8080: connect: connection refused
09:11:28 cache:service        [DEBUG] | creating 'http.failure-call.1A' data panic: found '2' errors (2)
- Get "http://localhost:8080/api/v1/ping": dial tcp 127.0.0.1:8080: connect: connection refused
- failed to connect to `host=localhost user= database=freqtrade`: dial error (dial tcp [::1]:5432: connect: cannot assign requested address) goroutine 1 [running]: main.defaultCommand(0xc000077720, 0x203000, 0xc00006bb40) 	/home/runner/work/fthelper/fthelper/metric/main.go:55 +0x81a github.com/kamontat/fthelper/shared/commandline/commands.(*Command).Start(0xc00006afc0, 0xc000077720, 0x2, 0xc00006bb40) /home/runner/work/fthelper/fthelper/shared/commandline/commands/command.go:27 +0x5a github.com/kamontat/fthelper/shared/commandline.(*cli).Start(0xc00002fa80, 0xc00001e070, 0x7, 0x7, 0xc0001ba708, 0xc0001c83f0) /home/runner/work/fthelper/fthelper/shared/commandline/cli.go:87 +0x9b2 main.main() /home/runner/work/fthelper/fthelper/metric/main.go:97 +0x477

09:13:29 map:merger           [DEBUG] | merging key=log from <nil> -> map[level:4]
09:13:29 map:merger           [DEBUG] | merging key=fs from <nil> -> map[env:map[fullpath:[/fth/.env.docker] mode:multiple type:auto]]
09:13:29 map:merger           [DEBUG] | merging key=flag from <nil> -> map[noenv:true]
09:13:29 map:merger           [DEBUG] | merging key=variables from <nil> -> map[current:/fth]
09:13:29 map:merger           [DEBUG] | merging key=config from <nil> -> map[fullpath:[/fth/configs] mode:multiple type:auto]
09:13:29 map:merger           [DEBUG] | merging key=clusters from <nil> -> []
09:13:29 map:merger           [DEBUG] | merging key=level from 4 -> 2
09:13:29 map:merger           [DEBUG] | merging key=fs from <nil> -> {"config":{"fullpath":["/fth/configs"],"mode":"multiple","type":"auto"},"env":{"fullpath":["/fth/.env.docker"],"mode":"multiple","type":"auto"}}
09:13:29 map:merger           [DEBUG] | merging key=variables from <nil> -> {"current":"/fth"}
09:13:29 map:merger           [DEBUG] | merging key=clusters from <nil> -> []
09:13:29 map:merger           [DEBUG] | merging key=internal from <nil> -> map[flag:{"noenv":true} log:map[level:2] meta:{"buildby":"goreleaser","commit":"9a789b82aae38ffde2a14d38e7097ce22cb44499","date":"2021-08-24T15:03:23Z","name":"ftmetric","version":"5.0.0-beta.12"}]
09:13:29 config:builder       [DEBUG] | base configuration is {"clusters":[""],"fs":{"config":{"fullpath":["/fth/configs"],"mode":"multiple","type":"auto"},"env":{"fullpath":["/fth/.env.docker"],"mode":"multiple","type":"auto"}},"internal":{"flag":{"noenv":true},"log":{"level":2},"meta":{"buildby":"goreleaser","commit":"9a789b82aae38ffde2a14d38e7097ce22cb44499","date":"2021-08-24T15:03:23Z","name":"ftmetric","version":"5.0.0-beta.12"}},"variables":{"current":"/fth"}}
09:13:29 fs:type:resolver     [DEBUG] | mark /fth/configs as directory
09:13:29 config:builder       [DEBUG] | loading config from wrapper of 'multiple': [/fth/configs (directory)]
09:13:29 map:merger           [DEBUG] | merging key=server from <nil> -> map[metrics-path:/metrics port:8090]
09:13:29 map:merger           [DEBUG] | merging key=warmup from <nil> -> map[enabled:true interval:30s]
09:13:29 map:merger           [DEBUG] | merging key=clusters from <nil> -> []
09:13:29 map:merger           [DEBUG] | merging key=freqtrade from <nil> -> map[cache:map[balance:3m count:1m locks:5m logs:1m performance:1m ping:1s profit:5m schedule_perf:1m show_config:1m stats:5m version:10m whitelist:15m] db:map[enabled:true name:freqtrade password: type:postgres url:localhost:5432 username:] http:map[apiver:v1 enabled:true password: query:map[logs:map[limit:100]] url:http://localhost:8080 username:]]
09:13:29 config:builder       [DEBUG] | found some config in env
09:13:29 config:builder       [DEBUG] | override 'cluster.1a.db.enabled' from '<**** [<nil>] -> 'f**** [bool] (env)
09:13:29 config:builder       [DEBUG] | override 'cluster.1a.http.url' from '<**** [<nil>] -> '"http****************81" [string] (env)
09:13:29 config:builder       [DEBUG] | override 'server.port' from '8*** [float64] -> '8*** [float64] (env)
09:13:29 config:builder       [DEBUG] | override 'clusters' from '[* [[]interface {}] -> '[*** [[]interface {}] (env)
09:13:29 config:builder       [DEBUG] | override 'cluster.1a.http.apiver' from '<**** [<nil>] -> 'v* [string] (env)
09:13:29 config:builder       [DEBUG] | override 'cluster.1a.http.username' from '<**** [<nil>] -> '"fr********" [string] (env)
09:13:29 config:builder       [DEBUG] | override 'cluster.1a.http.password' from '<**** [<nil>] -> '"s*****" [string] (env)
09:13:29 config:builder       [DEBUG] | override 'cluster.1a.http.enabled' from '<**** [<nil>] -> 't*** [bool] (env)
panic: found '2' errors (2)
- Get "http://localhost:8080/api/v1/ping": dial tcp 127.0.0.1:8080: connect: connection refused
- failed to connect to `host=localhost user= database=freqtrade`: dial error (dial tcp [::1]:5432: connect: cannot assign requested address)
goroutine 1 [running]:
main.defaultCommand(0xc0000b6c80, 0x203000, 0xc0000b8e80)
	/home/runner/work/fthelper/fthelper/metric/main.go:55 +0x81a
github.com/kamontat/fthelper/shared/commandline/commands.(*Command).Start(0xc0000b8300, 0xc0000b6c80, 0x2, 0xc0000b8e80)
	/home/runner/work/fthelper/fthelper/shared/commandline/commands/command.go:27 +0x5a
github.com/kamontat/fthelper/shared/commandline.(*cli).Start(0xc0000b0500, 0xc00001e070, 0x7, 0x7, 0xc00009a3c8, 0xc0000ae730)
	/home/runner/work/fthelper/fthelper/shared/commandline/cli.go:87 +0x9b2
main.main()

**Comment by frysee** [comment]
Well, that was a bit of a journey for someone who has no Go knowledge. Debugged into it with VS Code, the culprit seems to be the initial ping firing in Freqtrade::Initial(), added in 3d1c31f:
```
func (f *Freqtrade) Initial() error {
	var err1 = f.Connection.Http.Initial()
	var err2 = f.Connection.Db.Initial()
	return errors.New().And(err1, err2).Error()
}
```

What fails here is the toJson on the response:
`"found '1' errors (2)\n- json: Unmarshal(non-pointer maps.Mapper)\n"`

Removing err1 from the equation here fixes the issue and /metrics correctly exposes freqtrade metrics again.

**Comment by frysee** [comment]
Until fixed, forked images with the freqtrade ping on startup removed can be found here: 
https://github.com/frysee/fthelper/pkgs/container/ftmetric

**Comment by loewensteinph** [comment]
Thanks frysee that solves the isses, well partly still even though clusters are defined as following:

    environment:
      - FTH_FREQTRADE__DB__ENABLED=false
      - FTH_FREQTRADE__HTTP__URL=http://freqtrade1:8081
      - FTH_FREQTRADE__HTTP__USERNAME=freqtrader
      - FTH_FREQTRADE__HTTP__PASSWORD=secret
      - FTH_CLUSTERS=1A,2A,3A
      - FTH_CLUSTER__1A__DB__ENABLED=false      
      - FTH_CLUSTER__1A__HTTP__ENABLED=true
      - FTH_CLUSTER__1A__HTTP__URL=http://freqtrade1:8081
      - FTH_CLUSTER__1A__HTTP__APIVER=v1
      - FTH_CLUSTER__1A__HTTP__USERNAME=freqtrader
      - FTH_CLUSTER__1A__HTTP__PASSWORD=secret
      - FTH_CLUSTER__2A__DB__ENABLED=false      
      - FTH_CLUSTER__2A__HTTP__ENABLED=true
      - FTH_CLUSTER__2A__HTTP__URL=http://freqtrade2:8082
      - FTH_CLUSTER__2A__HTTP__APIVER=v1
      - FTH_CLUSTER__2A__HTTP__USERNAME=freqtrader
      - FTH_CLUSTER__2A__HTTP__PASSWORD=secret     
      - FTH_CLUSTER__3A__DB__ENABLED=false      
      - FTH_CLUSTER__3A__HTTP__ENABLED=true
      - FTH_CLUSTER__3A__HTTP__URL=http://freqtrade3:8083
      - FTH_CLUSTER__3A__HTTP__APIVER=v1
      - FTH_CLUSTER__3A__HTTP__USERNAME=freqtrader
      - FTH_CLUSTER__3A__HTTP__PASSWORD=secret          
      - FTH_SERVER__PORT=8091

All metrics are related to the instance specified as "FTH_FREQTRADE__HTTP__URL"

**Comment by frysee** [comment]
From a quick glance, you probably need to use 8080 on all of those since you are binding ftmetric to the **container port**, not the host port (assuming ftmetric is also spun up as a container).

The environment syntax also changed as far as I can tell from the changelog, you might have to use `FTC` for some of the keys. I am using config file which seems a lot cleaner when you need to configure a lot of things. Here is one that works for me (`--debug --no-env-file  --configs /fth/configs/common.json`):

```
{
  "clusters": ["live1", "live2", "live3"],
  "cluster": {},
  "freqtrade": {
    "http": {
      "enabled": true,
      "url": "http://freqtrade_dry1:8080",
      "apiver": "v1",
      "username": "freqtrade",
      "password": "freqtrade",
      "query": {
        "logs": {
          "limit": 100
        }
      }
    },
    "db": {
      "enabled": true,
      "type": "postgres",
      "url": "postgres:5432",
      "name": "freqtrade",
      "username": "freqtrade",
      "password": "freqtrade"
    },
    "cache": {
      "ping": "1s",
      "version": "10m",
      "balance": "3m",
      "count": "1m",
      "show_config": "1m",
      "profit": "5m",
      "locks": "5m",
      "whitelist": "15m",
      "performance": "1m",
      "logs": "1m",
      "stats": "5m",
      "schedule_perf": "1m"
    }
  },
  "server": {
    "port": 8090,
    "metrics-path": "/metrics"
  },
  "warmup": {
    "enabled": true,
    "interval": "30s"
  },
  "_": {
    "live1": {
      "freqtrade": {
        "http": {
          "url" : "http://freqtrade_live1:8080"
        },
        "db": {
          "name" : "freqtrade_live1"
        }
      }
    },
    "live2": {
      "freqtrade": {
        "http": {
          "url" : "http://freqtrade_live2:8080"
        },
        "db": {
          "name" : "freqtrade_live2"
        }
      }
    },
    "live3": {
      "freqtrade": {
        "http": {
          "url" : "http://freqtrade_live3:8080"
        },
        "db": {
          "name" : "freqtrade_live3"
        }
      }
    }
  }
}
```

**Comment by kamontat** [comment]
I think this issue contain too many issues now. Let me split to several problem that I found here..

**Comment by kamontat** [comment]
@frysee You right. v5 contains a lot of break change specially on cluster mode. I would suggest you to check on [MIGRATION PLAN](https://github.com/kamontat/fthelper/blob/main/metric/MIGRATION.md) first if you would like to upgrade to v5.

**Comment by frysee** [comment]
I think you can solve most of the issues by: 
- providing an up to date example for env/file configs, I found the current documentation a bit hard to grasp at the beginning.
- Fixing that startup ping check that is currently failing. (see above).

**Comment by kamontat** [comment]
Will this help? https://github.com/kamontat/fthelper/blob/main/metric/README.md#example

**Comment by frysee** [comment]
I think it would be nice to have a Json example with multiple freqtrade instances, apart from that: Yes. 

At some point it would be nice to have a bit of documentation on all possible parameters and their effect, but that's probably out of scope here.

**Comment by kamontat** [comment]
@loewensteinph I already found and fixed this issue.
Basically, I create connection from clusters list, so if cluster didn't exist no connection will be created.

Actually this issue you will spotted by no info connection was logged to console since no connection was created. Fixed `v5.0.0-beta.14`


For @frysee found issue, I already fixed as well. More detail on #29
